flocker/control/httpapi.py

```diff
@@ -664,6 +664,89 @@ def saved(_):
+        and restarted. This will also update the primary host of any attached
+        datasets.
+        """,
+        examples=[u"move container"],
+    )
+    @structured(
+        inputSchema={
+            '$ref':
+            '/v1/endpoints.json#/definitions/configuration_container_update',
+        },
+        outputSchema={
+            '$ref':
+            '/v1/endpoints.json#/definitions/configuration_container',
+        },
+        schema_store=SCHEMAS,
+    )
+    def update_containers_configuration(self, name, host):
+        """
+        Update the specified container's configuration.
+
+        :param unicode name: A unique identifier for the container within
+            the Flocker cluster.
+
+        :param unicode host: The address of the node on which the container
+            will run.
+
+        :return: An ``EndpointResponse`` describing the container which has
+            been updated.
+        """
+        deployment = self.persistence_service.get()
+
+        for node in deployment.nodes:
+            for application in node.applications:
+                if application.name == name:
+                    target_node = self._find_node_by_host(host, deployment)
+                    # We only need to perform a move if the node currently
+                    # hosting the container is not the node it's moving to.
+                    if node.hostname != host:
+                        # If the container has a volume, we need to add the
+                        # manifestation to the new host first.
+                        if application.volume is not None:
+                            dataset_id = application.volume.dataset.dataset_id
+                            target_node = target_node.transform(
+                                ("manifestations", dataset_id),
+                                application.volume.manifestation
+                            )
+                        # Now we can add the application to the new host.
+                        target_node = target_node.transform(
+                            ["applications"], lambda s: s.add(application))
+                        # And remove it from the current host.
+                        node = node.transform(
+                            ["applications"], lambda s: s.remove(application))
+                        # Finally we can now remove the manifestation from the
+                        # current host too.
+                        if application.volume is not None:
+                            dataset_id = application.volume.dataset.dataset_id
+                            node = node.transform(
+                                ("manifestations", dataset_id), discard
+                            )
+                        # Before updating the deployment and saving.
+                        deployment = deployment.update_node(node)
+                        deployment = deployment.update_node(target_node)
```

Comment:
I think that from `if node.hostname != host` to here would make a good method of `Deployment`.  `move_container` or something.  Do all this logic and return the new `Deployment` instance.  This method gets nicely shorter and the configuration mangling logic gets directly unit testable and re-usable (which is potentially beneficial to the API test suite).